source/extensions/matching/generic_inputs/environment/input.h

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include "envoy/matcher/matcher.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Matching {
+namespace GenericInputs {
+namespace Environment {
+
+class Input : public Envoy::Matcher::GenericDataInput {
+public:
+  explicit Input(absl::string_view name) {
+    // We read the env variable at construction time to avoid repeat lookups.
+    // This assumes that the environment remains stable during the process lifetime.
+    auto* value = getenv(name.data());
```

Comment:
Mostly curious what you think: if this side-effect was moved out of the constructor (into the factory perhaps?), this class would become a trivial wrapper around `absl::optional<std::string>`. This would make tests simpler, I think: `GenericTestInput` wouldn't be needed, and `TestGenericDataInputFactory` would only need to overload the method that retrieves env var with one that returns something expected in the test. Even though getenv() is pretty benign, side-effects in constructors always catch my attention...

Reply:
I like that, let me give that a go!